### ISSUE TITLE
How can I get the number of messages in Topic Partition ?

### ISSUE THREAD

**Issue by changemyminds** [confluentinc/confluent-kafka-dotnet#1944]
I want to get the number of messages in Topic Partition.

I see the [post](https://www.baeldung.com/java-kafka-count-topic-messages) which is introduced by java.

I use the Java code which can satisfy me.
```java
Properties properties = new Properties();
properties.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, BROKER_LIST);
properties.put(ConsumerConfig.GROUP_ID_CONFIG, GROUP_ID);
properties.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
properties.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
KafkaConsumer<String, String> consumer = new KafkaConsumer<>(properties);
List<TopicPartition> partitions = consumer.partitionsFor(TOPIC)
		.stream()
		.map(p -> new TopicPartition(TOPIC, p.partition()))
		.collect(Collectors.toList());
consumer.assign(partitions);
consumer.seekToEnd(Collections.emptySet());
Map<TopicPartition, Long> endPartitions = partitions.stream().collect(Collectors.toMap(Function.identity(), consumer::position));
long total = partitions.stream().mapToLong(endPartitions::get).sum();
System.out.println(total);
```

But I want to use **confluent-kafka-dotnet** to finish it so I convert Java Apache version to **C# confluent-kafka-dotnet** but it not working. The `total` value always `-1001` which is mean `Unset`. I'm so confused.
```csharp

[Fact]
public void Read_Kafka_Topic_NumberOfMessage()
{
	ConsumerConfig config = new ConsumerConfig
	{
		BootstrapServers = _bootstrapServers,
		GroupId = _groupId,
		AutoOffsetReset = AutoOffsetReset.Earliest,             
	};
	ConsumerBuilder<Null, string> builder = new ConsumerBuilder<Null, string>(config);
	builder.SetValueDeserializer(_kafkaConverter);
	IConsumer<Null, string> consumer = builder.Build();
	 
	List<TopicPartition> partitions = KafkaAdmin.GetTopicPartitions(_bootstrapServers, _topic);
	TopicPartition firstPartition = partitions.First();        
	consumer.Assign(partitions);
	consumer.Seek(new TopicPartitionOffset(firstPartition, Offset.End));

	Dictionary<TopicPartition, Offset> endPartitions = partitions.ToDictionary(k => k, v => consumer.Position(v));
	long total = partitions.Select(p => endPartitions[p].Value).Sum();
        System.Console.WriteLine(total); // -1001
}

public static class KafkaAdmin
{
	public static List<TopicPartition> GetTopicPartitions(string bootstrapServers, string topicValue)
	{
		AdminClientConfig adminClientConfig = new AdminClientConfig { BootstrapServers = bootstrapServers };
		using (var adminClient = new AdminClientBuilder(adminClientConfig).Build())
		{
			var meta = adminClient.GetMetadata(TimeSpan.FromSeconds(20));
			TopicMetadata? topicMetadata = meta.Topics.SingleOrDefault(t => topicValue.Equals(t.Topic));
			if (topicMetadata != null)
			{
				return topicMetadata.Partitions
					.Select(x => new TopicPartition(topicMetadata.Topic, x.PartitionId))
					.ToList();
			}
		}
		return new List<TopicPartition>();
	}
}

```
 
Am I using it wrong ?

**Comment by mhowlett** [comment]
suggest using `QueryWatermarkOffsets` - this will give you the low and high offsets, subtract to get number of partitions.

**Comment by edenhill** [comment]
number of messages.

But this is not exact, it will give you the maximum amount of messages between those two points, but it may be far less due to compaction, transaction control messages, etc.

TLDR: It is not possible to know the exact numbers of messages in a partition.
